src/lib/sticky-header/sticky-header.ts

```diff
@@ -0,0 +1,299 @@
+  stickyParent: HTMLElement | null;
+  /** The upper scrollable container. */
+  upperScrollableContainer: HTMLElement;
+  /**
+   * The original css of the sticky element, used to reset the sticky element
+   * when it is being unstuck
+   */
+  private _originalStyles = {} as CSSStyleDeclaration;
+  /**
+   * 'getBoundingClientRect().top' of CdkStickyRegion of current sticky header.
+   * It is used with '_stickyRegionBottomThreshold' to judge whether the current header
+   * need to be stuck.
+   */
+  private _stickyRegionTop: number;
+  /**
+   * Bottom of the sticky region offset by the height of the sticky header.
+   * Once the sticky header is scrolled to this position it will stay in place
+   * so that it will scroll naturally out of view with the rest of the sticky region.
+   */
+  private _stickyRegionBottomThreshold: number;
+
+  private _onScrollSubscription: Subscription;
+
+  private _onTouchSubscription: Subscription;
+
+  private _onResizeSubscription: Subscription;
+
+  constructor(element: ElementRef,
+              scrollable: Scrollable,
+              @Optional() public parentRegion: CdkStickyRegion,
+              platform: Platform) {
+    if (platform.isBrowser) {
+      this.element = element.nativeElement;
+      this.upperScrollableContainer = scrollable.getElementRef().nativeElement;
+      this._setStrategyAccordingToCompatibility();
+    }
+  }
+
+  ngAfterViewInit(): void {
+    if (!this._isPositionStickySupported) {
+
+      this.stickyParent = this.parentRegion != null ?
+        this.parentRegion._elementRef.nativeElement : this.element.parentElement;
+
+      let values = window.getComputedStyle(this.element, '');
+      this._originalStyles = {
+        position: values.position,
+        top: values.top,
+        right: values.right,
+        left: values.left,
+        bottom: values.bottom,
+        width: values.width,
+        zIndex: values.zIndex} as CSSStyleDeclaration;
+
+      this.attach();
+      this.defineRestrictionsAndStick();
+    }
+  }
+
+  ngOnDestroy(): void {
+    if (this._onScrollSubscription) {
+      this._onScrollSubscription.unsubscribe();
+    }
+
+    if (this._onResizeSubscription) {
+      this._onResizeSubscription.unsubscribe();
+    }
+
+    if (this._onTouchSubscription) {
+      this._onTouchSubscription.unsubscribe();
+    }
+  }
+
+  /**
+   * Check if current browser supports sticky positioning. If yes, apply
+   * sticky positioning. If not, use the original implementation.
+   */
+  private _setStrategyAccordingToCompatibility(): void {
+    this._isPositionStickySupported = isPositionStickySupported();
+    if (this._isPositionStickySupported) {
+      this.element.style.top = '0px';
+      this.element.style.cssText += 'position: -webkit-sticky; position: sticky; ';
+      // TODO add css class with both 'sticky' and '-webkit-sticky' on position
+      // when @Directory supports adding CSS class
+    }
+  }
+
+  attach() {
```

Comment:
:+1: Changed to `private _attachEventListeners`, and Added a description  "Add listeners for events that affect sticky positioning."